File: src/lib/sunday-tools/common/race.ts
import { Region, RegionList } from '../shared/region';
import { createParser } from '../skills/parser/ConditionParser';
import { Strategy } from '../runner/definitions';
import { Rule30CARng } from '../shared/random';
import { GameHpPolicy } from '../health/game.policy';
import { NoopHpPolicy } from '../health/health-policy';
import { StrategyHelpers } from '../runner/runner.types';
import { registerAllDynamicConditions } from '../full-sim/register-all';
import { RaceEventBus } from './race-events';
import { Runner } from './runner';
import type { CreateRunner } from './runner';
import type { PRNG } from '../shared/random';
import type { IStrategy } from '../runner/definitions';
import type { DefaultParser } from '../skills/parser/definitions';
import type {
  CourseData,
  IGrade,
  IGroundCondition,
  ISeason,
  ITimeOfDay,
  IWeather,
} from '../course/definitions';

export type RunnerMap = Map<number, Runner>;

export type RunnerSnapshot = {
  position: number;
  currentLane: number;
  currentSpeed: number;
};

export type SimulationSettings = {
  mode: 'compare' | 'normal';
  /**
   * Whether the simulation should account for the health system
   */
  healthSystem: boolean;
  /**
   * Whether the simulation should account for the section modifier
   */
  sectionModifier: boolean;
  /**
   * Whether the simulation should account for the rushed state
   */
  rushed: boolean;
  /**
   * Whether the simulation should account for the downhill mode
   */
  downhill: boolean;
  /**
   * Whether the simulation should account for the spot struggle
   */
  spotStruggle: boolean;
  /**
   * Whether the simulation should account for the dueling
   */
  dueling: boolean;
  /**
   * Whether the simulation should account for wit checks
   *
   * Values:
   * - true: Wit checks are enabled
   * - false: Wit checks are disabled so skills always pass.
   */
  witChecks: boolean;
  /**
   * Position keep simulation mode.
   *
   * Values:
   * - 0: position keep off
   * - 2: virtual position keep
   */
  positionKeepMode: number;
  /**
   * Optional per-skill stamina drain overrides.
   * Key: base skill ID, value: drain fraction (0 to 1).
   */
  staminaDrainOverrides?: Record<string, number>;
};

export interface RaceLifecycleObserver {
  /** Called once when prepareRound completes (before any ticks) */
  onRoundStart(race: Race, seed: number): void;

  /** Called before the main runner update loop each frame */
  onBeforeTick(race: Race, dt: number): void;

  /** Called after each individual runner is updated in a frame */
  onAfterRunnerTick(race: Race, runner: Runner, dt: number): void;

  /** Called the moment a runner crosses the finish line */
  onRunnerFinished(race: Race, runner: Runner): void;

  /** Called when all runners have finished the round */
  onRoundEnd(race: Race): void;
}

export type DuelingRates = {
  runaway: number;
  frontRunner: number;
  paceChaser: number;
  lateSurger: number;
  endCloser: number;
};

export type RaceParameters = {
  ground: IGroundCondition;
  weather: IWeather;
  season: ISeason;
  timeOfDay: ITimeOfDay;
  grade: IGrade;
  strategyCounts?: Map<IStrategy, number>;
  commonSkills?: Map<string, number>;
  numUmas?: number;

  [key: string]: any;
};

export type RaceSimulatorProps = {
  /**
   * The parameters for the race simulation
   */
  parameters: RaceParameters;
  /**
   * The course data for the race
   */
  course: CourseData;

  /**
   * The settings for the race simulation
   */
  settings: SimulationSettings;

  /**
   * The number of skill samples to use for the race simulation
   */
  skillSamples: number;

  duelingRates: DuelingRates;
};

export class Race {
  // ===================
  // Private
  // ===================

  public seed!: number;
  public rng!: PRNG;
  public lastRunnerId!: number;
  public settings: SimulationSettings;
  public duelingRates: DuelingRates;
  public events: RaceEventBus;

  // ===================
  // Public
  // ===================

  declare public pacer: Runner;
  declare public commonSkills: Map<string, number>;
  declare public strategyCounts: Map<IStrategy, number>;
  declare public runnerOrder: Map<number, number>;
  declare public previousRunnerOrder: Map<number, number>;
  declare public runnersPerStrategy: Map<IStrategy, Array<Runner>>;
  declare public accumulatedTime: number;

  declare public parser: DefaultParser;
  declare public wholeCourse: RegionList;
  declare public skillSamples: number;
  declare public roundIteration: number;

  public course: CourseData;
  public ground: IGroundCondition;
  public weather: IWeather;
  public season: ISeason;
  public timeOfDay: ITimeOfDay;
  public grade: IGrade;

  public runners: Map<number, Runner>;
  public finishedRunners: Array<number>;
  public runnerSnapshots: Map<number, RunnerSnapshot>;

  constructor(props: RaceSimulatorProps & { collector?: RaceLifecycleObserver }) {
    this.course = props.course;

    this.ground = props.parameters.ground;
    this.weather = props.parameters.weather;
    this.season = props.parameters.season;
    this.timeOfDay = props.parameters.timeOfDay;
    this.grade = props.parameters.grade;

    this.runners = new Map();
    this.finishedRunners = [];
    this.runnerSnapshots = new Map();

    this.settings = props.settings;
    this.duelingRates = props.duelingRates;
    this.events = new RaceEventBus();

    // Temporary backward compatibility while callsites move to event subscriptions.
    const observer = props.collector;
    if (observer) {
      this.events.on('round-start', (race, seed) => observer.onRoundStart(race, seed));
      this.events.on('before-tick', (race, dt) => observer.onBeforeTick(race, dt));
      this.events.on('after-runner-tick', (race, runner, dt) =>
        observer.onAfterRunnerTick(race, runner, dt),
      );
      this.events.on('runner-finished', (race, runner) => observer.onRunnerFinished(race, runner));
      this.events.on('round-end', (race) => observer.onRoundEnd(race));
    }
  }

  public onInitialize(): void {
    registerAllDynamicConditions();

    // Default values
    this.seed = -1;
    this.lastRunnerId = 0;

    this.parser = createParser();
    this.wholeCourse = new RegionList();
    this.wholeCourse.push(new Region(0, this.course.distance));
    this.skillSamples = 0;
    this.roundIteration = 0;
  }

  public onPrepare(): void {
    this.accumulatedTime = 0;

    this.commonSkills = new Map();
    this.strategyCounts = new Map();
    this.runnerOrder = new Map();
    this.previousRunnerOrder = new Map();
    this.runnersPerStrategy = new Map();
    this.runnerSnapshots.clear();
  }

  // ===================
  // Getters
  // ===================

  public get baseSpeed(): number {
    return 20.0 - (this.course.distance - 2000) / 1000.0;
  }

  // ==================
  // Lifecycle
  // ==================

  /**
   * Assign unique gate positions to all runners.
   *
   * Uses a Fisher-Yates shuffle seeded from the race RNG to produce
   * a fair, deterministic, unique gate assignment for each runner.
   *
   * Gate numbers are 0-8 (9 positions), mapped to lane positions
   * by the runner using `gate * course.horseLane`.
   */
  private assignGates(): void {
    if (!this.rng) throw new Error('Race RNG not set');

    // Create gate pool (0 to 8, always 9 gates)
    const gates = Array.from({ length: 9 }, (_, i) => i);

    // Fisher-Yates shuffle using race RNG
    for (let i = gates.length - 1; i > 0; i--) {
      const j = this.rng.uniform(i + 1);
      [gates[i], gates[j]] = [gates[j], gates[i]];
    }

    // Assign first N gates to runners (in iteration order, which is insertion order for Map)
    let gateIndex = 0;
    for (const runner of this.runners.values()) {
      runner.setGate(gates[gateIndex]);
      gateIndex++;
    }
  }

  public validateRaceSetup(): void {
    if (this.runners.size === 0) throw new Error('No runners added to race');
  }

  public prepareRace() {
    const commonSkills = new Map<string, number>();
    const strategyCounts = new Map<IStrategy, number>([
      [Strategy.Runaway, 0],
      [Strategy.FrontRunner, 0],
      [Strategy.PaceChaser, 0],
      [Strategy.LateSurger, 0],
      [Strategy.EndCloser, 0],
    ]);

    const runnersPerStrategy = new Map<IStrategy, Array<Runner>>([
      [Strategy.Runaway, []],
      [Strategy.FrontRunner, []],
      [Strategy.PaceChaser, []],
      [Strategy.LateSurger, []],
      [Strategy.EndCloser, []],
    ]);

    for (const runner of this.runners.values()) {
      const strategy = runner.strategy;
      const skills = runner.skillIds;

      for (const skill of skills) {
        const count = commonSkills.get(skill) ?? 0;
        commonSkills.set(skill, count + 1);
      }

      const stratCount = strategyCounts.get(strategy)!;
      strategyCounts.set(strategy, stratCount + 1);

      const runners = runnersPerStrategy.get(strategy)!;
      runners.push(runner);
      runnersPerStrategy.set(strategy, runners);
    }

    this.commonSkills = commonSkills;
    this.strategyCounts = strategyCounts;
    this.runnersPerStrategy = runnersPerStrategy;
    this.runnerOrder = new Map();
    this.previousRunnerOrder = new Map();

    this.accumulatedTime = 0;
    this.roundIteration = 0;

    return this;
  }

  public prepareRound(masterSeed: number) {
    this.seed = masterSeed;
    this.rng = new Rule30CARng(masterSeed);

    this.assignGates();

    for (const runner of this.runners.values()) {
      // Generate runner-specific RNG
      const runnerRng = new Rule30CARng(this.rng.int32());

      // Setup health policy
      if (this.settings.healthSystem) {
        // Create health policy RNG (separate from runner's main RNG)
        const hpRng = new Rule30CARng(this.rng.int32());
        const healthPolicy = new GameHpPolicy(this.course, this.ground, hpRng);
        runner.setHealthPolicy(healthPolicy);
      } else {
        runner.setHealthPolicy(NoopHpPolicy);
      }

      runner.onPrepare(runnerRng);
    }

    // Increment the round iteration as setup is complete, so this value should not be used until the next round.
    this.roundIteration++;

    this.events.emit('round-start', this, masterSeed);
  }

  /**
   * Steps the race simulation for a given time step.
   *
   * @param dt The time step to advance the simulation by.
   * @param runners The runners to step.
   */
  public onUpdate(dt: number): void {
    this.events.emit('before-tick', this, dt);

    this.accumulatedTime += dt;

    if (this.settings.mode === 'normal') {
      const sortedRunners = Array.from(this.runners.values())
        .filter((runner) => !this.finishedRunners.includes(runner.id))
        .sort((a, b) => b.position - a.position);

      this.previousRunnerOrder = new Map(this.runnerOrder);
      this.runnerOrder.clear();
      for (let i = 0; i < sortedRunners.length; i += 1) {
        this.runnerOrder.set(sortedRunners[i].id, i + 1);
      }
    }

    this.runnerSnapshots.clear();
    for (const runner of this.runners.values()) {
      if (this.finishedRunners.includes(runner.id)) continue;
      this.runnerSnapshots.set(runner.id, {
        position: runner.position,
        currentLane: runner.currentLane,
        currentSpeed: runner.currentSpeed,
      });
    }

    // Internally set the pacer.
    // ? Should this be done every frame?
    const pacer = this.getPacer();
    if (pacer) {
      this.pacer = pacer;
    }

    for (const runner of this.runners.values()) {
      if (this.finishedRunners.includes(runner.id)) continue;

      runner.onUpdate(dt);
      this.events.emit('after-runner-tick', this, runner, dt);

      if (runner.finished) {
        this.finishedRunners.push(runner.id);
        this.events.emit('runner-finished', this, runner);
      }
    }
  }

  /**
   * Runs the race simulation
   *
   * This will run until all runners have finished the race
   */
  public run() {
    // Reset the runners to step.
    this.finishedRunners = [];

    // Run the race until all runners have finished.
    while (this.runners.size !== this.finishedRunners.length) {
      this.onUpdate(1 / 15);
    }

    this.events.emit('round-end', this);
  }

  private getPacer() {
    // Select furthest-forward front runner
    for (const strategy of [Strategy.Runaway, Strategy.FrontRunner]) {
      // ! NOTE: might revise this.
      const frontRunners = Array.from(this.runners.values()).filter(
        (runner) => runner.positionKeepStrategy === strategy,
      );

      const firstRunner = frontRunners[0];

      if (frontRunners.length > 0) {
        const pacer = frontRunners.reduce((max, currRunner) => {
          return currRunner.position > max.position ? currRunner : max;
        }, firstRunner);

        return pacer;
      }
    }

    // Get pacerOverride uma
    const pacerOverrideUma = this.pacer;

    if (pacerOverrideUma) {
      return pacerOverrideUma;
    }

    // Otherwise, lucky pace (set pacerOverride)
    for (const strategy of [Strategy.PaceChaser, Strategy.LateSurger, Strategy.EndCloser]) {
      // ! NOTE: might revise this.
      const runnersWithStrat = Array.from(this.runners.values()).filter((runner) =>
        StrategyHelpers.strategyMatches(runner.positionKeepStrategy, strategy),
      );

      const firstRunner = runnersWithStrat[0];

      if (runnersWithStrat.length > 0) {
        const luckyPacer = runnersWithStrat.reduce((max, currRunner) => {
          return currRunner.position > max.position ? currRunner : max;
        }, firstRunner);

        luckyPacer.positionKeepStrategy = Strategy.FrontRunner;

        return luckyPacer;
      }
    }

    // Otherwise, get virtual pacemaker
    // (this should never happen though)
    const pacer = this.pacer;

    if (pacer) {
      pacer.positionKeepStrategy = Strategy.FrontRunner;
      return pacer;
    }

    return null;
  }

  // === runner management ===

  /**
   * Register a new runner to the race
   */
  public addRunner(params: CreateRunner) {
    const runner = Runner.create(this, this.lastRunnerId, params);
    this.runners.set(this.lastRunnerId, runner);

    this.lastRunnerId++;

    return this;
  }

  // === Settings ===

  public setRushed(rushed: boolean) {
    this.settings.rushed = rushed;
    return this;
  }

  public setDownhill(downhill: boolean) {
    this.settings.downhill = downhill;
    return this;
  }

  public setSpotStruggle(spotStruggle: boolean) {
    this.settings.spotStruggle = spotStruggle;
    return this;
  }

  public setDueling(dueling: boolean) {
    this.settings.dueling = dueling;
    return this;
  }

  public setWitChecks(witChecks: boolean) {
    this.settings.witChecks = witChecks;
    return this;
  }
}

File: packages/uma-sim-vacuum/src/race.rs
//! The `Race` **aggregate root** (port of `common/race.ts`).
//!
//! Owns the field of [`Runner`]s and is the only thing that mutates them. The
//! tick loop is **snapshot-based** (see `.plans/.../context.md`): each frame
//! builds one immutable [`FieldSnapshot`] from the *current* state, then steps
//! every runner against that same frozen snapshot, so resolution order is
//! irrelevant and there is no intra-frame state drift.
//!
//! Runners are stored in a `Vec<Runner>` indexed by their `RunnerId` (assigned
//! in insertion order) — deterministic iteration without the overhead of a map,
//! and contiguous for the slice-based pacing / coordinator services.

use std::collections::HashMap;

use uma_sim_primitives::course::model::CourseData;
use uma_sim_primitives::events::{RaceObservation, RaceObserver, RaceObservers};
use uma_sim_primitives::position_keep::{update_position_keep_coefficient, PositionKeepContext};
use uma_sim_primitives::race_support::{
    build_field_snapshot, build_field_view, condition_value, FieldOrderTracker,
};
use uma_sim_primitives::runner::lifecycle::{CreateRunner, PrepareContext};
use uma_sim_primitives::runner::mechanics::DuelingRates;
use uma_sim_primitives::runner::physics::{
    update_first_position_in_late_race, DuelingInput, FieldInputs, SkillTriggerInputs,
    UpdateContext,
};
use uma_sim_primitives::runner::skills::FieldView;
use uma_sim_primitives::runner::Runner;
use uma_sim_primitives::shared_kernel::ids::RunnerId;
use uma_sim_primitives::shared_kernel::language::{GroundCondition, Strategy};
use uma_sim_primitives::shared_kernel::params::RaceParameters;
use uma_sim_primitives::shared_kernel::region::{Region, RegionList};
use uma_sim_primitives::shared_kernel::rng::{Prng, Xoshiro256StarStar};
use uma_sim_primitives::skills::condition::catalog::build_catalog;
use uma_sim_primitives::skills::condition::dynamic::register_all_dynamic_conditions;
use uma_sim_primitives::skills::condition::language::ConditionParser;
use uma_sim_primitives::skills::condition::{ConditionCatalog, ConditionResolution};
use uma_sim_primitives::stamina::game_policy::GameStaminaPolicy;
use uma_sim_primitives::stamina::policy::{NoopStaminaPolicy, StaminaPolicy};

/// Frame duration (15 FPS).
const FRAME_DT: f64 = 1.0 / 15.0;

/// Toggles and tuning that configure a simulation run.
#[derive(Debug, Clone)]
pub struct SimulationSettings {
    /// Whether the game stamina policy (HP) is active.
    pub health_system: bool,
    /// Whether per-section wisdom variance is applied (currently always on).
    pub section_modifier: bool,
    /// Whether the rushed (temptation) mechanic is enabled.
    pub rushed: bool,
    /// Whether downhill mode is enabled.
    pub downhill: bool,
    /// Whether spot-struggle is enabled.
    pub spot_struggle: bool,
    /// Whether dueling is enabled.
    pub dueling: bool,
    /// Whether wit checks gate skill activation.
    pub wit_checks: bool,
    /// Position-keep mode (`2` enables virtual position keeping).
    pub position_keep_mode: i32,
    /// Number of activation samples drawn per skill trigger.
    pub skill_samples: usize,
    /// Per-base-skill recovery (stamina-drain) override modifiers.
    pub stamina_drain_overrides: HashMap<String, f64>,
}

impl Default for SimulationSettings {
    fn default() -> Self {
        SimulationSettings {
            health_system: true,
            section_modifier: true,
            rushed: true,
            downhill: true,
            spot_struggle: true,
            dueling: true,
            wit_checks: true,
            position_keep_mode: 2,
            skill_samples: 1,
            stamina_drain_overrides: HashMap::new(),
        }
    }
}

/// The race aggregate root.
pub struct Race {
    /// The course being raced.
    pub course: CourseData,
    /// Ground condition.
    pub ground: GroundCondition,
    /// Simulation settings.
    pub settings: SimulationSettings,
    /// Race-wide parameters (augmented with field composition at prepare time).
    pub race_params: RaceParameters,
    /// Per-strategy dueling rates (compare-mode artificial dueling).
    pub dueling_rates: Option<DuelingRates>,

    /// The field, indexed by `RunnerId`.
    runners: Vec<Runner>,
    /// Ids of finished runners, in finish order (append-only).
    finished_runners: Vec<RunnerId>,

    /// The static condition catalog (owns the parser's backing data).
    catalog: ConditionCatalog,
    /// The whole course as a region list.
    whole_course: RegionList,
    /// Round index (selects which sampled trigger fires).
    round_iteration: usize,
    /// Elapsed race time in seconds.
    accumulated_time: f64,
    /// Master seed of the current round.
    seed: u64,
    /// The race RNG (drives gate assignment + per-runner sub-streams).
    rng: Box<dyn Prng>,

    /// Pacer + finishing-order state, threaded across frames.
    order_tracker: FieldOrderTracker,
    /// Per-strategy counts.
    strategy_counts: HashMap<Strategy, u32>,
    /// Common (shared) skills across the field.
    common_skills: HashMap<String, u32>,

    /// Lifecycle observers.
    observers: RaceObservers,
}

impl RaceObservation for Race {
    fn course_distance(&self) -> f64 {
        self.course.distance
    }
    fn pacer_position(&self) -> Option<f64> {
        self.order_tracker.pacer_position
    }
    fn seed(&self) -> u64 {
        self.seed
    }
    fn accumulated_time(&self) -> f64 {
        self.accumulated_time
    }
}

impl Race {
    /// Build a race for a course/ground with the given settings + parameters.
    ///
    /// Registers the dynamic condition catalog once and builds the static
    /// catalog + whole-course regions.
    pub fn new(
        course: CourseData,
        ground: GroundCondition,
        settings: SimulationSettings,
        race_params: RaceParameters,
        dueling_rates: Option<DuelingRates>,
    ) -> Self {
        register_all_dynamic_conditions();
        let mut whole_course = RegionList::new();
        whole_course.push(Region::new(0.0, course.distance));
        Race {
            ground,
            settings,
            race_params,
            dueling_rates,
            runners: Vec::new(),
            finished_runners: Vec::new(),
            catalog: build_catalog(),
            whole_course,
            round_iteration: 0,
            accumulated_time: 0.0,
            seed: 0,
            rng: Box::new(Xoshiro256StarStar::from_u64_seed(0)),
            order_tracker: FieldOrderTracker::new(),
            strategy_counts: HashMap::new(),
            common_skills: HashMap::new(),
            observers: RaceObservers::new(),
            course,
        }
    }

    /// Course base speed: `20 - (distance - 2000) / 1000`.
    pub fn base_speed(&self) -> f64 {
        20.0 - (self.course.distance - 2000.0) / 1000.0
    }

    /// Number of runners in the field.
    pub fn runner_count(&self) -> usize {
        self.runners.len()
    }

    /// Read-only access to the field.
    pub fn runners(&self) -> &[Runner] {
        &self.runners
    }

    /// Finish order (runner ids).
    pub fn finished_runners(&self) -> &[RunnerId] {
        &self.finished_runners
    }

    /// Register a lifecycle observer.
    pub fn subscribe(&mut self, observer: Box<dyn RaceObserver>) {
        self.observers.subscribe(observer);
    }

    /// Add a runner to the field (assigns the next `RunnerId`).
    pub fn add_runner(&mut self, props: CreateRunner) -> RunnerId {
        let id = RunnerId(self.runners.len() as u32);
        let runner = Runner::create(
            id,
            &self.course,
            self.ground,
            props,
            Box::new(NoopStaminaPolicy),
            Box::new(Xoshiro256StarStar::from_u32_seed(id.0)),
        );
        self.runners.push(runner);
        id
    }

    /// Prepare the field for a round: count field composition, assign gates,
    /// spawn per-runner RNGs + stamina policies, and reset every runner.
    pub fn prepare_round(&mut self, master_seed: u64) {
        self.accumulated_time = 0.0;
        self.finished_runners.clear();
        self.order_tracker.reset();

        self.prepare_race();

        self.seed = master_seed;
        self.rng = Box::new(Xoshiro256StarStar::from_u64_seed(master_seed));

        let base_speed = self.base_speed();
        let parser = ConditionParser::new(&self.catalog);

        // Fisher-Yates gate shuffle (9 gates) seeded from the race RNG.
        let mut gates: Vec<i64> = (0..9).collect();
        for i in (1..gates.len()).rev() {
            let j = self.rng.uniform(i as u32 + 1) as usize;
            gates.swap(i, j);
        }

        let mut runners = std::mem::take(&mut self.runners);
        for (idx, runner) in runners.iter_mut().enumerate() {
            runner.gate = gates[idx];
            let runner_rng: Box<dyn Prng> =
                Box::new(Xoshiro256StarStar::from_u32_seed(self.rng.int32()));
            let policy: Box<dyn StaminaPolicy> = if self.settings.health_system {
                let hp_rng: Box<dyn Prng> =
                    Box::new(Xoshiro256StarStar::from_u32_seed(self.rng.int32()));
                Box::new(GameStaminaPolicy::new(&self.course, self.ground, hp_rng))
            } else {
                Box::new(NoopStaminaPolicy)
            };
            runner.health_policy = policy;
            runner.wit_checks_enabled = self.settings.wit_checks;
            runner.dueling_enabled = self.settings.dueling;
            runner.spot_struggle_enabled = self.settings.spot_struggle;
            runner.downhill_enabled = self.settings.downhill;
            runner
                .stamina_drain_overrides
                .clone_from(&self.settings.stamina_drain_overrides);

            let ctx = PrepareContext {
                course: &self.course,
                base_speed,
                // Synthetic engine: static approximate regions + the ×10
                // position-keep window (ADR-0005, no paradigm branch).
                condition_resolution: ConditionResolution::Static,
                pos_keep_end_multiplier: 10.0,
                race_params: &self.race_params,
                whole_course: &self.whole_course,
                parser: &parser,
                skill_samples: self.settings.skill_samples,
                round_iteration: self.round_iteration,
            };
            runner.on_prepare(runner_rng, &ctx);
        }
        self.runners = runners;

        self.round_iteration += 1;
        self.emit_round_start(master_seed);
    }

    /// Reset per-round race params for the vacuum field: the synthetic engine
    /// does not fold in field composition (no live field to count), so the
    /// strategy-count / common-skill aggregates are cleared.
    fn prepare_race(&mut self) {
        self.strategy_counts = HashMap::new();
        self.common_skills = HashMap::new();

        self.race_params.num_umas = Some(self.runners.len() as u32);
        self.race_params.strategy_counts = None;
        self.race_params.common_skills = None;
    }

    /// Run the race to completion (every runner finished).
    pub fn run(&mut self) {
        while self.finished_runners.len() < self.runners.len() {
            self.on_update(FRAME_DT);
        }
        self.emit_round_end();
    }

    /// Advance the whole field one `dt`-second step (snapshot-based).
    pub fn on_update(&mut self, dt: f64) {
        self.emit_before_tick(dt);
        self.accumulated_time += dt;

        let snapshot = build_field_snapshot(
            &mut self.runners,
            &self.finished_runners,
            &mut self.order_tracker,
        );

        let base_speed = self.base_speed();
        let mut runners = std::mem::take(&mut self.runners);
        for runner in &mut runners {
            if self.finished_runners.contains(&runner.id) {
                continue;
            }
            update_position_keep_coefficient(runner);
            let field = build_field_view(runner.id, &snapshot);
            let position_keep = PositionKeepContext {
                position_keep_mode: self.settings.position_keep_mode,
                num_runners: snapshot.num_active as usize,
                pacer_position: snapshot.pacer_position,
                pacer_is_self: snapshot.pacer == Some(runner.id),
                second_place_position: snapshot.second_place_position,
            };
            // The aggregate is the *producer*: it resolves every field-presence
            // input from **approximate condition values + synthetic dueling
            // rates** (ADR-0005 synthetic producer) and hands the step a pure
            // `FieldInputs` plus a field-free course context. There is no
            // paradigm branch and no live field to read.
            let field_inputs =
                resolve_field_inputs(runner, &field, self.dueling_rates, position_keep);
            let ctx = UpdateContext {
                base_speed,
                accumulated_time: self.accumulated_time,
                course: &self.course,
                downhill_enabled: self.settings.downhill,
            };
            runner.on_update(dt, &field_inputs, &ctx);
        }
        self.runners = runners;

        // No cross-runner contention coordinator in the vacuum engine: there is
        // no live field, so proximity dueling / spot-struggle never emerge.

        update_first_position_in_late_race(&mut self.runners, self.course.distance);

        self.emit_runner_ticks_and_finishes(dt);
    }

    fn emit_runner_ticks_and_finishes(&mut self, dt: f64) {
        let mut observers = std::mem::take(&mut self.observers);
        let mut newly_finished: Vec<RunnerId> = Vec::new();
        for runner in &self.runners {
            if self.finished_runners.contains(&runner.id) {
                continue;
            }
            observers.emit_after_runner_tick(self, runner, dt);
            if runner.finished {
                newly_finished.push(runner.id);
            }
        }
        for id in newly_finished {
            self.finished_runners.push(id);
            if let Some(runner) = self.runners.iter().find(|r| r.id == id) {
                observers.emit_runner_finished(self, runner);
            }
        }
        self.observers = observers;
    }

    fn emit_round_start(&mut self, seed: u64) {
        let mut observers = std::mem::take(&mut self.observers);
        observers.emit_round_start(self, seed);
        self.observers = observers;
    }

    fn emit_before_tick(&mut self, dt: f64) {
        let mut observers = std::mem::take(&mut self.observers);
        observers.emit_before_tick(self, dt);
        self.observers = observers;
    }

    fn emit_round_end(&mut self) {
        let mut observers = std::mem::take(&mut self.observers);
        observers.emit_round_end(self);
        self.observers = observers;
    }
}

/// Resolve the [`FieldInputs`] for one runner this tick, from **approximate
/// condition values + synthetic dueling rates** (ADR-0005 synthetic producer).
///
/// There is no live field: side-block / overtake are read from the runner's
/// pre-resolved approximate conditions, and dueling is the artificial roll
/// driven by the per-strategy [`DuelingRates`] (which may be absent). The step
/// never sees how these were produced.
fn resolve_field_inputs<'a>(
    runner: &Runner,
    field: &'a FieldView,
    dueling_rates: Option<DuelingRates>,
    position_keep: PositionKeepContext,
) -> FieldInputs<'a> {
    FieldInputs {
        side_blocked: condition_value(runner, "blocked_side") == 1,
        overtaking: condition_value(runner, "overtake") == 1,
        dueling: DuelingInput::Artificial(dueling_rates),
        position_keep,
        skill_triggers: SkillTriggerInputs { field },
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashMap;
    use uma_sim_primitives::runner::lifecycle::RunnerAptitudes;
    use uma_sim_primitives::runner::test_support::{test_course, test_race_params};
    use uma_sim_primitives::shared_kernel::language::{Aptitude, Mood};
    use uma_sim_primitives::shared_kernel::params::StatLine;

    fn props(name: &str, strategy: Strategy) -> CreateRunner {
        CreateRunner {
            outfit_id: "100302".to_owned(),
            name: name.to_owned(),
            mood: Mood::Normal,
            strategy,
            aptitudes: RunnerAptitudes {
                distance: Aptitude::A,
                strategy: Aptitude::A,
                surface: Aptitude::A,
            },
            stats: StatLine {
                speed: 1000,
                stamina: 1000,
                power: 1000,
                guts: 1000,
                wit: 800,
            },
            skills: vec![],
            forced_positions: HashMap::new(),
            injected_debuffs: vec![],
            forced_rushed_regions: vec![],
            forced_dueling_regions: vec![],
            forced_spot_struggle_regions: vec![],
            forced_rank: vec![],
        }
    }

    fn race_with(n: u32) -> Race {
        let mut race = Race::new(
            test_course(),
            GroundCondition::Firm,
            SimulationSettings::default(),
            test_race_params(),
            None,
        );
        let strategies = [
            Strategy::Runaway,
            Strategy::FrontRunner,
            Strategy::PaceChaser,
            Strategy::LateSurger,
            Strategy::EndCloser,
        ];
        for i in 0..n {
            let s = strategies[(i as usize) % strategies.len()];
            race.add_runner(props(&format!("R{i}"), s));
        }
        race
    }

    #[test]
    fn nine_runner_race_finishes_all_deterministically() {
        let mut race = race_with(9);
        race.prepare_round(12345);
        race.run();
        assert_eq!(race.finished_runners().len(), 9);
        // Every runner crossed the line.
        assert!(race.runners().iter().all(|r| r.finished));
        // Finish times are positive and ordered by finish sequence.
        let first = race.finished_runners()[0];
        let winner = race
            .runners()
            .iter()
            .find(|r| r.id == first)
            .expect("winner present");
        assert!(winner.finish_time > 0.0);
    }

    #[test]
    fn same_seed_same_finish_order() {
        let mut a = race_with(9);
        a.prepare_round(777);
        a.run();
        let mut b = race_with(9);
        b.prepare_round(777);
        b.run();
        assert_eq!(a.finished_runners(), b.finished_runners());
    }

    #[test]
    fn different_seed_may_differ_but_all_finish() {
        let mut race = race_with(9);
        race.prepare_round(2024);
        race.run();
        assert_eq!(race.finished_runners().len(), 9);
        // Gates are a unique permutation of 0..9.
        let mut gates: Vec<i64> = race.runners().iter().map(|r| r.gate).collect();
        gates.sort_unstable();
        gates.dedup();
        assert_eq!(gates.len(), 9);
    }
}
